client/go/outline/vpn/tun_linux.go

```diff
@@ -36,23 +50,74 @@ func newTUNDevice(name string) (io.ReadWriteCloser, error) {
 	if err != nil {
 		return nil, err
 	}
+	defer func() {
+		if err != nil {
+			tun.Interface.Close()
+		}
+	}()
+
 	if tun.Name() != name {
-		return nil, fmt.Errorf("tun device name mismatch: requested `%s`, created `%s`", name, tun.Name())
+		return nil, fmt.Errorf("TUN device name mismatch: requested `%s`, created `%s`", name, tun.Name())
+	}
+
+	// Wait for the TUN device to be available in NetworkManager
+	for retries := tunAPIRetryCount; retries > 0; retries-- {
+		slog.Debug("trying to locate TUN device in NetworkManager...", "tun", name)
+		tun.nmDev, err = nm.GetDeviceByIpIface(name)
+		if tun.nmDev != nil && err == nil {
+			break
+		}
+		slog.Debug("waiting for TUN device to be available in NetworkManager", "err", err)
+		time.Sleep(tunAPIRetryDelay)
+	}
+	if tun.nmDev == nil {
+		return nil, fmt.Errorf("failed to locate the TUN device `%s` in NetworkManager", tun.Name())
+	}
+	slog.Debug("found TUN device in NetworkManager", "dev", tun.nmDev.GetPath())
+
+	// Let NetworkManager take care of the TUN device
+	if err = setTUNDeviceManaged(tun.nmDev, true); err != nil {
+		return nil, err
 	}
+
+	slog.Info("TUN device successfully created", "name", tun.Name(), "dev", tun.nmDev.GetPath())
 	return tun, nil
 }
 
-// waitForTUNDeviceToBeAvailable waits for the TUN device with the given name to be available
-// in the specific NetworkManager.
-func waitForTUNDeviceToBeAvailable(nm gonm.NetworkManager, name string) (dev gonm.Device, err error) {
-	for retries := 20; retries > 0; retries-- {
-		slog.Debug("trying to find tun device in NetworkManager...", "tun", name)
-		dev, err = nm.GetDeviceByIpIface(name)
-		if dev != nil && err == nil {
-			return
+func (tun *tunDevice) Close() (err error) {
+	tun.Interface.Close()
+	if err = deleteTUNDevice(tun.nm, tun.Name()); err == nil {
+		slog.Info("TUN device deleted", "name", tun.Name())
+	}
+	return
+}
+
+func setTUNDeviceManaged(dev gonm.Device, value bool) error {
+	for retries := tunAPIRetryCount; retries > 0; retries-- {
+		if err := dev.SetPropertyManaged(value); err != nil {
```

Comment:
Do you really need to call set multiple times?
Or should you just wait for the value to be reflected on the Get call?

Reply:
Updated, set once and wait to confirm for multiple times.